### DISCUSSION TITLE
How to implement a contain inside the 'or' operator?

### DISCUSSION THREAD

**Discussion by emilianoyaryura** [supabase/supabase#19129] [Questions] [answered]
How are you? I'm having trouble implementing the ```or``` operator in supabase.

I need to create a 'for you' tab, so I'm trying to filter the posts depending on the tags and people the user follows.

When I implement the ```in``` and ```contains``` operators separately, it works fine. But when I implement the ```or``` operator, it fails. Also, I've tried doing ```.or(`tags.contains.${tags || []}, created_by.id.in.${following || []}`)```

This works okay.
```
  const { data, error } = await supabase
    .from("post")
    .select(
      "title, description, likes, cover_image, views, slug, created_at, published, id, tags, created_by!inner( name, profile_image, default_profile_img_color, id, description, username, cafecito )"
    )
    .eq("published", true)
    .contains("tags", tags || [])
    .order("created_at", { ascending: false })
    .limit(limit);
```

This also works okay
```
  const { data, error } = await supabase
    .from("post")
    .select(
      "title, description, likes, cover_image, views, slug, created_at, published, id, tags, created_by!inner( name, profile_image, default_profile_img_color, id, description, username, cafecito )"
    )
    .eq("published", true)
    .in('created_by.id', following || [])
    .order("created_at", { ascending: false })
    .limit(limit);
```

This doesn't work
```
  const { data, error } = await supabase
    .from("post")
    .select(
      "title, description, likes, cover_image, views, slug, created_at, published, id, tags, created_by!inner( name, profile_image, default_profile_img_color, id, description, username, cafecito )"
    )
    .eq("published", true)
    .or(`tags.cn.${tags || []}, created_by.id.in.${following || []}`)
    .order("created_at", { ascending: false })
    .limit(limit);
```

**Comment by GaryAustin1** [comment, marked as answer]
The .or() requires PostgREST format for filters.
The example here happens to show the format for in and contains....
https://supabase.com/docs/reference/javascript/or
![image](https://github.com/supabase/supabase/assets/54564956/a39c072a-2dd7-4e42-b692-50d832e6687b)

All the details are here:
https://postgrest.org/en/stable/references/api/tables_views.html#operators

**Reply by emilianoyaryura** [reply]
thanks for answering!

If `tags` and `following` are arrays, how should I do it? Because I've tried changing the `cn` for `cs` as you showed me, but it continues returning a _"failed to parse logic tree"_ error.

Thank you

**Reply by GaryAustin1** [reply]
I used this once and added it to the docs, but the example went away with v2 docs...
`.or(`id.in.(${arrList}),arraycol.cs.{${arr}}`)	`
Note the () for a list and {} for an array in Postgres syntax.

**Reply by emilianoyaryura** [reply]
Sorry for bothering you, but I can't replicate that. This is my api route. It works fine, except when I add the `or` operator.

```
import { NextApiRequest, NextApiResponse } from "next";
import { supabase } from "../../../lib/supabase-client";

async function handler(req: NextApiRequest, res: NextApiResponse) {
  if (req.method !== "POST") {
    // Return an error response for unsupported request methods
    res.status(405).json({ error: "Method Not Allowed" });
    return;
  }

  const {
    tags,
    limit,
    following,
  }: {
    tags: string[] | null;
    limit: number;
    following: string[] | null;
  } = req.body;

  if ((!tags || tags.length === 0) && (!following || following.length === 0)) {
    res.status(200).json({ posts: [] });
  }

  const { data, error } = await supabase
    .from("post")
    .select(
      "title, description, likes, cover_image, views, slug, created_at, published, id, tags, created_by!inner( name, profile_image, default_profile_img_color, id, description, username, cafecito )"
    )
    .eq("published", true)
    .or(`tags.cs.{${tags}}, created_by.id.in.(${following})`)
    .order("created_at", { ascending: false })
    .limit(limit);

  if (error) {
    res.status(500).json({ error: error.message });
  }

  res.status(200).json({ posts: data });
}

export default handler;
```

This returns a "failed to parse logic tree" error.

Do you know what Im doing wrong? Thanks

pd: the arrays columns are called `tags` and `following`

**Reply by GaryAustin1** [reply]
`created_by.id.in  `  Not sure what this is.  I assume you wanted just `created_by.in`

**Reply by emilianoyaryura** [reply]
thanks! I was trying to access the id, as created_by is a reference field. Sorry for wasting your time and thank you very much for your help.

**Reply by GaryAustin1** [reply]
So is this resolved?  Thanks.
